src/main/java/rx/schedulers/JavaFxScheduler.java

```diff
@@ -18,194 +18,151 @@
+		return new JavaFxWorker();
+	}
+
+	/**
+	 * A Worker implementation which manages a queue of QueuedRunnable for execution on the Java FX Application thread
+	 * For a simpler implementation the queue always contains at least one element.
+	 * {@link #head} is the element, which is in execution or was last executed
+	 * {@link #tail} is an atomic reference to the last element in the queue, or null when the worker was disposed
+	 * Recursive actions are not preferred and inserted at the tail of the queue as any other action would be
+	 * The Worker will only schedule a single job with {@link Platform#runLater(Runnable)} for when the queue was previously empty
+	 */
+	private static class JavaFxWorker extends Worker implements Runnable {
+		private volatile QueuedRunnable                  head = new QueuedRunnable(null); /// only advanced in run(), initialised with a starter element
+		private final    AtomicReference<QueuedRunnable> tail = new AtomicReference<>(head); /// points to the last element, null when disposed
+
+		private static class QueuedRunnable extends AtomicReference<QueuedRunnable> implements Subscription, Action0 {
+			private volatile Action0 action;
+
+			private QueuedRunnable(Action0 action) {
+				this.action = action;
+			}
+
+			@Override
+			public void unsubscribe() {
+				action = null;
+			}
+
+			@Override
+			public boolean isUnsubscribed() {
+				return action == null;
+			}
+
+			@Override
+			public void call() {
+				Action0 action = this.action;
+				if (action != null) {
+					action.call();
+				}
+				this.action = null;
+			}
+		}
+
+		@Override
+		public void unsubscribe() {
+			tail.set(null);
+			QueuedRunnable qr = this.head;
+			while (qr != null) {
+				qr.unsubscribe();
+				qr = qr.getAndSet(null);
+			}
+		}
+
+		@Override
+		public boolean isUnsubscribed() {
+			return tail.get() == null;
+		}
+
+		@Override
+		public Subscription schedule(final Action0 action, long delayTime, TimeUnit unit) {
+			long delay = Math.max(0, unit.toMillis(delayTime));
+			assertThatTheDelayIsValidForTheJavaFxTimer(delay);
+
+			final QueuedRunnable queuedRunnable = new QueuedRunnable(action);
+			if (delay == 0) { // delay is too small for the java fx timer, schedule it without delay
+				return schedule(queuedRunnable);
+			}
+
+			final Timeline timer = new Timeline(new KeyFrame(Duration.millis(delay), event -> schedule(queuedRunnable)));
+			timer.play();
+
+			return Subscriptions.create(() -> {
+				queuedRunnable.unsubscribe();
+				timer.stop();
+			});
+		}
+
+		@Override
+		public Subscription schedule(final Action0 action) {
+			if (isUnsubscribed()) {
+				return Subscriptions.unsubscribed();
+			}
+
+			final QueuedRunnable queuedRunnable = action instanceof QueuedRunnable ? (QueuedRunnable) action : new QueuedRunnable(action);
+
+			QueuedRunnable tailPivot;
+			do {
+				tailPivot = tail.get();
+			} while (tailPivot != null && !tailPivot.compareAndSet(null, queuedRunnable));
```

Comment:
This loop appends elements and is the only critical operation for the queue.
It's a simple, atomic, linked list.
The current tail element is fetched from the pointer `tail` and remembered as `tailPivot` for later evaluation.
Iff `tailPivot==null` the worker was disposed/unsubscribed and we bail out (see line 146 as well).
Otherwise we try to append the new `queuedRunnable` to the `tailPivot`, which we expect to have no successor and thus do `CAS` with null. This fails iff some other thread successfully appended its element. In this case we start over.
When we successfully appended our element to the queue it is available for execution, but we need to advance `tail` to point to the new element as all other threads will be busy waiting for the correct `tailPivot`. This is done in line 148. The `CAS` there will fail iff the worker was disposed/unsubscribed which means `tail` has been cleared and all waiting threads will bail out anyways.